pynobo.py

```diff
@@ -145,115 +142,191 @@ class API:
             while discovered_hubs:
                 (discover_ip, discover_serial) = discovered_hubs.pop()
                 try:
-                    self.connect_hub(discover_ip, discover_serial)
+                    await self.connect_hub(discover_ip, discover_serial)
                     break  # We connect to the first valid hub, no reason to try the rest
                 except Exception as e:
                     # This might not have been the Ecohub we wanted (wrong serial)
                     self.logger.warning("Could not connect to {}:{} - {}".format(
-                                        discover_ip, discover_serial, e))
+                        discover_ip, discover_serial, e))
         else:
             # check if we have an IP
-            if not ip:
+            if not self.ip:
                 self.logger.error('Could not connect, no ip address provided')
                 raise ValueError('Could not connect, no ip address provided')
 
             # check if we have a valid serial before we start connection
-            if len(serial) != 12:
+            if len(self.serial) != 12:
                 self.logger.error('Could not connect, no valid serial number provided')
                 raise ValueError('Could not connect, no valid serial number provided')
 
-            self.connect_hub(ip, serial)
+            await self.connect_hub(self.ip, self.serial)
 
-        if not self.client:
+        if not self.writer:
             self.logger.error("Failed to connect to Ecohub")
             raise Exception("Failed to connect to Ecohub")
 
-        # start thread for continously receiving new data from hub
-        self.socket_receive_thread = threading.Thread(target=self.socket_receive)
-        self.socket_receive_thread.daemon = True
-        self.socket_receive_exit_flag = threading.Event()
-        self.socket_received_all_info = threading.Event()
-        self.socket_connected = threading.Event()
-        self.socket_connected.set()
-        self.socket_receive_thread.start()
-
-        # Fetch all info
-        self.send_command([self.API.GET_ALL_INFO])
-        self.socket_received_all_info.wait()
-
-        self.logger.info('connected to Nobø Hub')
-
-
-    def connect_hub(self, ip, serial):
-        """Attempt initial connection and handshake
+        self.logger.info('connected to Nobø Ecohub')
+
+    async def stop(self):
+        """Stop the keep-alive and receiver tasks and close the connection to Nobø Ecohub."""
+        self.keep_alive_task.cancel()
+        with suppress(asyncio.CancelledError):
+            await self.keep_alive_task
+        self.socket_receive_task.cancel()
+        with suppress(asyncio.CancelledError):
+            await self.keep_alive_task
+        await self.close()
+        self.logger.info('disconnected from Nobø Ecohub')
+
+    async def close(self):
+        """Close the connection to Nobø Ecohub."""
+        if self.writer:
```

Comment:
I think we should also close `self.reader`.

Reply:
I thought so to, but surprisingly there is no close method: https://docs.python.org/3/library/asyncio-stream.html#streamreader

Closing the writer closes the underlying socket: https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamWriter.close